File: apps/web/src/lib/trustSafety/chatSafetyPolicy.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import {
  CHAT_SAFETY_POLICY_VERSION,
  reviewChatMessageSafety,
} from "./chatSafetyPolicy.ts";

const acceptedMessages = [
  "Would you consider trading this for my Charizard?",
  "That is a sick card with killer artwork.",
  "Misty's Determination is one of my favorite cards.",
  "Would you take $40 for the card?",
  "I can find you a cleaner copy in my vault.",
  "Here is the Grookai card link: https://grookai.com/cards/GV-PK-BASE-001",
];

const blockedMessages = [
  ["Email me at collector@example.com", "personal_contact_information"],
  ["Text me at 303-555-0118", "personal_contact_information"],
  ["Open https://example.com/card", "external_link"],
  ["Message me on Discord", "off_platform_contact"],
  ["Pay me with a gift card", "unsafe_payment_request"],
  ["I will hurt you", "direct_threat"],
  ["You are an idiot", "targeted_harassment"],
  ["Send me nudes", "sexual_solicitation"],
  ["I do not want to live", "self_harm_risk"],
  ["aaaaaaaaaaaa", "spam_pattern"],
];

test("chat safety policy accepts normal collector language", () => {
  for (const message of acceptedMessages) {
    const decision = reviewChatMessageSafety(message);
    assert.equal(decision.allowed, true, message);
    assert.equal(decision.policyVersion, CHAT_SAFETY_POLICY_VERSION);
  }
});

test("chat safety policy blocks concrete safety risks with stable reason codes", () => {
  for (const [message, expectedReason] of blockedMessages) {
    const decision = reviewChatMessageSafety(message);
    assert.equal(decision.allowed, false, message);
    assert.equal(decision.reasonCode, expectedReason, message);
    assert.ok(decision.userMessage, message);
  }
});

test("empty and oversized messages fail closed", () => {
  assert.equal(reviewChatMessageSafety("   ").reasonCode, "empty_message");
  assert.equal(
    reviewChatMessageSafety("a".repeat(2001)).reasonCode,
    "message_too_long",
  );
});

File: test/chat_safety_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:grookai_vault/services/network/chat_safety_policy.dart';

void main() {
  group('Chat Safety Policy V1', () {
    test('accepts normal collector language', () {
      const messages = [
        'Would you consider trading this for my Charizard?',
        'That is a sick card with killer artwork.',
        "Misty's Determination is one of my favorite cards.",
        'Would you take \$40 for the card?',
        'I can find you a cleaner copy in my vault.',
        'Here is the Grookai card link: https://grookai.com/cards/GV-PK-BASE-001',
      ];

      for (final message in messages) {
        expect(
          reviewChatMessageSafety(message).allowed,
          isTrue,
          reason: message,
        );
      }
    });

    test('blocks concrete safety risks with stable reason codes', () {
      final cases = <String, ChatSafetyReasonCode>{
        'Email me at collector@example.com':
            ChatSafetyReasonCode.personalContactInformation,
        'Text me at 303-555-0118':
            ChatSafetyReasonCode.personalContactInformation,
        'Open https://example.com/card': ChatSafetyReasonCode.externalLink,
        'Message me on Discord': ChatSafetyReasonCode.offPlatformContact,
        'Pay me with a gift card': ChatSafetyReasonCode.unsafePaymentRequest,
        'I will hurt you': ChatSafetyReasonCode.directThreat,
        'You are an idiot': ChatSafetyReasonCode.targetedHarassment,
        'Send me nudes': ChatSafetyReasonCode.sexualSolicitation,
        'I do not want to live': ChatSafetyReasonCode.selfHarmRisk,
        'aaaaaaaaaaaa': ChatSafetyReasonCode.spamPattern,
      };

      for (final entry in cases.entries) {
        final decision = reviewChatMessageSafety(entry.key);
        expect(decision.allowed, isFalse, reason: entry.key);
        expect(decision.reasonCode, entry.value, reason: entry.key);
        expect(decision.userMessage, isNotEmpty, reason: entry.key);
      }
    });

    test('empty and oversized messages fail closed', () {
      expect(
        reviewChatMessageSafety('   ').reasonCode,
        ChatSafetyReasonCode.emptyMessage,
      );
      expect(
        reviewChatMessageSafety('a' * 2001).reasonCode,
        ChatSafetyReasonCode.messageTooLong,
      );
    });
  });
}
